File: apps/cargo-tracker/src/Cargotracker/Estimation/Domain/Model/RouteCandidate.hs
{- | 経路候補 エンティティ (US01, IT2)

Estimate 集約配下のエンティティ。1 件の見積に対して 0 件以上の RouteCandidate
を `rank` 昇順 (0 = 最優先) で保持する。

- rank: 0..N の整数 (集約内でユニーク、0 を直行便とする慣例)
- transitDays: 所要日数 (1 以上)
- estimatedCost: 概算料金 (0 以上、IT2 では円単位 Int で簡略化)
- voyageNumbers: 経由航海番号テキスト (1 件以上)。
  cross-BC ACL を尊重し Routing.VoyageNumber 値オブジェクトを直接
  保持しない (T-06 Rule 4)。Application 層で文字列変換する。
-}
module Cargotracker.Estimation.Domain.Model.RouteCandidate
  ( RouteCandidate (..),
    mkRouteCandidate,
  ) where

import Data.Text (Text)
import qualified Data.Text as T

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data RouteCandidate = RouteCandidate
  { rank :: !Int
  , transitDays :: !Int
  , estimatedCost :: !Int
  , voyageNumbers :: ![Text]
  }
  deriving stock (Eq, Show)

mkRouteCandidate ::
  Int -> Int -> Int -> [Text] -> Either DomainError RouteCandidate
mkRouteCandidate r td cost vns
  | r < 0 = Left (InvalidBookingId "rank must be >= 0")
  | td < 1 = Left (InvalidBookingId "transitDays must be >= 1")
  | cost < 0 = Left (InvalidBookingId "estimatedCost must be >= 0")
  | null vns = Left (InvalidBookingId "voyageNumbers must not be empty")
  | any (T.null . T.strip) vns =
      Left (InvalidBookingId "voyageNumbers must not contain empty strings")
  | otherwise =
      Right
        RouteCandidate
          { rank = r
          , transitDays = td
          , estimatedCost = cost
          , voyageNumbers = map T.strip vns
          }

File: docs/article/source/ruby/apps/cargo-tracker/packs/estimation/app/domain/estimation/domain/route_candidate.rb
# frozen_string_literal: true

module Estimation
  module Domain
    # 経路候補（RouteCandidate）。見積に紐づく不変値オブジェクト（US01）。
    # 経由港・所要日数・概算料金・航海番号を保持する。Routing の一時候補を Estimation へ永続化する
    # （ADR-0004・IT7 で統合）。
    class RouteCandidate
      attr_reader :voyage_number, :transit_port, :transit_days, :estimated_cost, :rank

      def initialize(voyage_number:, transit_days:, estimated_cost:, transit_port: nil, rank: 0)
        raise ArgumentError, "航海番号は必須です" if voyage_number.to_s.strip.empty?
        raise ArgumentError, "所要日数は正である必要があります" if transit_days.to_i <= 0

        cost = estimated_cost.is_a?(BigDecimal) ? estimated_cost : BigDecimal(estimated_cost.to_s)
        raise ArgumentError, "概算料金は正である必要があります" if cost <= 0

        @voyage_number = voyage_number
        @transit_port = transit_port
        @transit_days = transit_days.to_i
        @estimated_cost = cost
        @rank = rank.to_i
        freeze
      end

      def ==(other)
        other.is_a?(RouteCandidate) && other.voyage_number == voyage_number &&
          other.transit_days == transit_days && other.estimated_cost == estimated_cost
      end
      alias eql? ==
      def hash = [ voyage_number, transit_days, estimated_cost ].hash
    end
  end
end
